### ISSUE TITLE
JpegEncoder producing incorrect output on linux-arm (since Beta006)

### ISSUE THREAD

**Issue by qdraw** [SixLabors/ImageSharp#871]
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am running the latest version of ImageSharp
- [x] I have verified if the problem exist in both `DEBUG` and `RELEASE` mode
- [x] I have searched [open](https://github.com/SixLabors/ImageSharp/issues) and [closed](https://github.com/SixLabors/ImageSharp/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported
- [x] I have discussed this on gitter

### Description

I have some Image encoding issues on linux-arm. This code bellow works with `beta-0005`, but not `beta-0006` or newer. I have also tested it with the latest nightly build `1.0.0-dev002459` The issues are only visible on linux-arm, on other platforms it is working well 

#### UPDATE (22/05/2019, @antonfirsov)
Resizing is not relevant here, it's the issue of the jpeg encoder.

#### TL;DR:
- load jpeg -> resize -> save jpeg [PINK]
- load jpeg -> save jpeg [PINK]
- load png -> save jpeg [PINK]
- load gif -> save jpeg [PINK]
- load jpeg -> resize -> save png [OK]
- load jpeg -> save png [OK]

The images are looking like this, I have tried it with multiple jpeg images
![QZNU6ZQ3I6C7WV76TUB2X7C57E](https://user-images.githubusercontent.com/1826996/55347543-73324980-54b5-11e9-9c1b-7bb2ba2ed6b3.jpg)
Source of the image:
https://raw.githubusercontent.com/SixLabors/ImageSharp/master/tests/Images/Input/Jpg/baseline/Lake.jpg

### Steps to Reproduce

I'm initially used it for resizing but it happens also when you write the image:
I'm using the following code:
```cs
using (var inputStream = _iStorage.ReadStream(subPath))
using (var image = Image.Load(inputStream))
{
	image.Save(outputStream, new JpegEncoder{
		Quality = 100,
	});
						
	_iStorage.WriteStream(outputStream, outputHash);  
}

public Stream ReadStream(string path, int maxRead = -1)
{
	if ( ! ExistFile(path) ) throw new FileNotFoundException(path);
	
	FileStream fileStream;
	fileStream = new FileStream(path, FileMode.Open, FileAccess.Read,FileShare.Read, maxRead, false);
	return fileStream;
}

public bool WriteStream(Stream stream, string path)
{
	if ( !stream.CanRead ) return false;
	
	stream.Seek(0, SeekOrigin.Begin);
				
	using (var fileStream = new FileStream(path, FileMode.Create, FileAccess.Write))
	{
		stream.CopyTo(fileStream);
	}
	stream.Dispose();
	return true;
}
```

### System Configuration
When running on linux-arm as release and debug

- ImageSharp version: 0006 and newer
- Other ImageSharp packages and versions:
- Environment (Operating system, version and so on): linux-arm
- .NET Framework version: build as: 2.2.104, runtime: 2.1.8
- Additional information: running on a Raspberry Pi 2 or 3

-- 
Edit: tested based on the feedback of @JimBobSquarePants:, this produces a 'normal' image  
```cs
image.Save(outputStream, new PngEncoder{
    ColorType = PngColorType.Rgb, 
   CompressionLevel = 9, 
});
```

**Comment by antonfirsov** [comment]
info from gitter discussions: PNG -> Jpeg also fails, which indicates an issue with JpegEncoder

**Comment by qdraw** [comment]
I have tried some things based on the release notes: (will update this comment if i have tried something)
[x] commit: 8f3658da6c0040198c01dbf892a9ea77d0556407 ()  PR: #804 - pink

**Comment by JimBobSquarePants** [comment]
#804 is all decoder work which we know is fine so the change must have taken place between #784 and there.

**Comment by antonfirsov** [comment]
We are getting close! However it's very strange, that no PR [between](https://github.com/SixLabors/ImageSharp/pulls?page=2&q=is%3Apr+is%3Aclosed) 784 and 804 touches anything related to Jpeg encoding. What is the state after [merging](https://github.com/SixLabors/ImageSharp/commit/441942b771232f3b1f4c4a54bdf79f48bc1ca6bc) #801 ?

**Comment by JimBobSquarePants** [comment]
I’m smelling a jit issue here...

**Comment by qdraw** [comment]
The state after #801 is a pink image

**Comment by antonfirsov** [comment]
Only 6 PR-s left to binary search ... if we did no mistake so far. Are you super sure, that #784 is OK?
If yes, let's point to the middle .. #789?

**Comment by qdraw** [comment]
I have #784 checked again, and I was wrong this one is also pink. I have updated the comment above

**Comment by antonfirsov** [comment]
Another idea:

What if you wire out the pooling memory allocator? `Configuration.Default.MemoryAllocator = new SimpleGcMemoryAllocator()`

Should I provide a detailed plan to track down the breaking PR? (Or are you alredy doing it? :) )

**Comment by qdraw** [comment]
@antonfirsov I'm currently checking PR's using a very simple application that only convert the image ('Lake.jpg' from the examples) to a image on device. Do you have suggestions which PR I should track?

**Comment by qdraw** [comment]
> What if you wire out the pooling memory allocator? Configuration.Default.MemoryAllocator = new SimpleGcMemoryAllocator()
What should I do to test this?

**Comment by antonfirsov** [comment]
> What should I do to test this?

Same thing: pink/not pink. (Just make sure you do it before running other code.)

Re tracking down the PR-s:
It should be between 724-801 now. Try pointing to the middle (binary search FTW).

**Comment by antonfirsov** [comment]
A search tree that might be useful:
```
            #727 
        #729 
    #731 
            #742
        #744
#746 
            #747 
        #751
    #768
        #771
            #783
```
PR-s that are more to the left, should be checked first. PR-s which leave JpegEncoder unaffected (in theory) are not listed. You should be able to track down the breaking PR in `<=4` further checks!

**Comment by qdraw** [comment]
@antonfirsov thanks for the list, I will binary search the commits and update this comment

```
            #727 -  ccfabf8fe7ccdb6c7cd8ca13b5fa153365f39ec2 - normal image
        #729
    #731
            #742
        #744
#746
            #747
        #751
    #768
        #771
            #783
```

**Comment by qdraw** [comment]
@antonfirsov
Q1: what is the value of Vector.IsHardwareAccelerated on your device?
```
Vector.IsHardwareAccelerated ->
False
```
Q2: [EDITED] what if you undefine SUPPORTS_EXTENDED_INTRINSICS in ImageSharp.csproj?
I can't find this value. How does this work?

**Comment by antonfirsov** [comment]
@qdraw open `ImageSharp.csproj` with a text editor (or in VS: right click -> edit csproj), and delete [this PropertyGroup](https://github.com/SixLabors/ImageSharp/blob/master/src/ImageSharp/ImageSharp.csproj#L31).

**Comment by qdraw** [comment]
@antonfirsov Ah I see https://github.com/SixLabors/ImageSharp/commit/f52be32abafa7d164d9c31f2ce4851576c667ac9#diff-e8ae3d9cc22f149b899610bb6fe4a5cb, lets get back to the latest master. This created the same pink image

**Comment by qdraw** [comment]
commit: 0dedf86b0b403c2da244b40993ae749f9d36b297 is [PINK]
commit: d4be172dcc44f19136e286cce5b855985a58ffba is [PINK]

**Comment by qdraw** [comment]
The combination when loading Rgb24 images with has been changed in: `b14c35ffafe2f762f0ee19cbca34982d638ed039` I have tested this on the latest master build (2019-04-14)

__The client__
`using (var image = Image.Load<Rgb24>(inputStream))`

__The change reverted__
```cs

        public Rgb24 Rgb
        {
            // If this is changed to ShortMethod then several jpeg encoding tests fail
            // on 32 bit Net 4.6.2 and NET 4.7.1
            [MethodImpl(InliningOptions.ShortMethod)]
            get => Unsafe.As<Rgba32, Rgb24>(ref this);

            [MethodImpl(InliningOptions.ShortMethod)]
            set => Unsafe.As<Rgba32, Rgb24>(ref this) = value;
        }
```

However this is not working on with the default client parameter 
`using (var inputStream = ReadStream(inputPath))` -> Pink image
and the comments are describing issues with .NET Framework and x86-Windows

**Comment by brianpopow** [comment]
As mentioned in gitter, i was able to reproduce this issue with docker. I have create a branch in my fork, so others can reproduce that to: [Issue871](https://github.com/brianpopow/ImageSharp/tree/feature/Issue871)

This adds an additional Project ImageSharp.DockerSandbox which will load and save the images from #871 and #914. To execute this with docker just use following commands in the root directory of ImageSharp:

```
docker-compose build
docker-compose up
```

This will execute the ImageSharp.DockerSandbox project inside docker with a arm32 runtime image.
The output images will be in `docker_output` folder.

Changing the `Rgb24` property in  Rgba32.cs seems to fix this issue:

```
public Rgb24 Rgb
{
	get => new Rgb24(this.R, this.G, this.B);

	set
	{
		this.R = value.R;
		this.G = value.G;
		this.B = value.B;
	}
}
```

**Comment by brianpopow** [comment]
@qdraw it seems you are able to run tests on a real device. Would it be possible for you to run my changes from [pull/915](https://github.com/SixLabors/ImageSharp/pull/915) on a real hardware? It would be great to verify, that this change indeed fixes the issue.

**Comment by dlemstra** [comment]
I wonder if this can also be reproduced in a super small sample so we can report this as a bug upstream.

**Comment by JimBobSquarePants** [comment]
Would be wise to do it, raising an issue at least would make them aware.

**Comment by dontspeak** [comment]
Nice that there is progress with this issue. I tried to debug into the issue on my rasperry, but as of now I did not got it managed to create a valid linux-arm compile with visual studio using the ImageSharp source code.

Is this issue completely fixed or do we only have a workaround? If the root cause really only was the RGB24 Contructur, why then we not has´ve these issues all over ImageSharp, but only when encoding JPEGs.

Is there still testing needed? I can support.

**Comment by JimBobSquarePants** [comment]
It _should_ be absolutely fixed. 

It's not the `Rgb24` constructor that is the issue at all but rather how the jitter on Linux ARM handles the use of `Unsafe.As` on a non-inlined property when used in combination with bulk operations. It only affected the jpeg encoder because that was the only area in which we call `Rgba32` to `Rgb24` in that manner.

>Is there still testing needed? I can support.

If you could download the latest build from MyGet and test that we'd really appreciate it. 👍

**Comment by dontspeak** [comment]
MyGet Package works perfect now on my Raspberry. Thanks!

**Comment by JimBobSquarePants** [comment]
Hooray!
